src/container/srv_container.c

```diff
@@ -5158,30 +5182,55 @@ cont_op_with_svc(struct ds_pool_hdl *pool_hdl, struct cont_svc *svc,
 	else
 		ABT_rwlock_wrlock(svc->cs_lock);
 
+	/* TODO: add client-provided metadata RPC key, lookup in cs_ops KVS, assign dup_op */
+
 	switch (opc) {
 	case CONT_CREATE:
-		rc = cont_create(&tx, pool_hdl, svc, rpc);
+		rc = cont_create(&tx, pool_hdl, svc, dup_op, rpc);
 		if (likely(rc == 0)) {
 			metrics = pool_hdl->sph_pool->sp_metrics[DAOS_CONT_MODULE];
 			d_tm_inc_counter(metrics->create_total, 1);
 		}
+		if (dup_op)
+			goto out_lock;
+
 		break;
 	case CONT_OPEN_BYLABEL:
 		olbl_in = crt_req_get(rpc);
 		olbl_out = crt_reply_get(rpc);
 		rc = cont_lookup_bylabel(&tx, svc, olbl_in->coli_label, &cont);
+		/* TODO: idempotent rc=0 return for dup_op case. */
 		if (rc != 0)
 			goto out_lock;
 		/* NB: call common cont_op_with_cont() same as CONT_OPEN case */
 		rc = cont_op_with_cont(&tx, pool_hdl, cont, rpc, &update_mtime, cont_proto_ver);
 		uuid_copy(olbl_out->colo_uuid, cont->c_uuid);
 		break;
 	case CONT_DESTROY_BYLABEL:
-		dlbl_in = crt_req_get(rpc);
-		rc = cont_lookup_bylabel(&tx, svc, dlbl_in->cdli_label, &cont);
-		if (rc != 0)
+	case CONT_DESTROY:
+		if (opc == CONT_DESTROY_BYLABEL) {
+			dlbl_in = crt_req_get(rpc);
+			clbl    = dlbl_in->cdli_label;
+			rc      = cont_lookup_bylabel(&tx, svc, dlbl_in->cdli_label, &cont);
+		} else {
+			uuid_unparse(in->ci_uuid, cuuid);
+			rc = cont_lookup(&tx, svc, in->ci_uuid, &cont);
+		}
+		if (rc == -DER_NONEXIST && dup_op) {
```

Comment:
[Just a comment] If the request is a dup, I imagine we will directly generate a reply based on the info in the `ops` KVS, and not execute into the core request handling logic (i.e., not enter this `switch` statement, and certainly not attempting to look up the container).

One thing might need some thoughts: If the request is not a dup, and the container is not found (`rc == -DER_NONEXIST`), then we need to put `-DER_NONEXIST` into the `ops` record of this operation. Do we need to differentiating errors that are conclusive results of operations with errors such as `-DER_NOMEM`, `-DER_GRPVER`, `-DER_TIMEDOUT`?